### PR TITLE
Incentive splitter

### PR DESCRIPTION
**Pull request by abhishek-k-sharma-lab49**
Splitter Implementation done

### CODE REVIEW

File: contracts/incentive/ISplitter.sol [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.5.0 <0.9.0;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
```
Comment (amandeeprehal-lab49) [inline contracts/incentive/ISplitter.sol:4]: Are you using this?

File: contracts/incentive/IVault.sol [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.5.0 <0.9.0;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
```
Comment (amandeeprehal-lab49) [inline contracts/incentive/IVault.sol:4]: Are you using this?

File: contracts/incentive/Splitter.sol [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    function _totalWeightForAllVaults() private returns (uint256 totalWeight) {
+        for (uint256 i = 0; i < _vaults.length; i++) {
+            IVault tempVault = _vaults[i];
+            uint256 weight = _weightForVault(tempVault);
+            totalWeight += weight;
+        }
+    }
+
+    function _weightForVault(IVault vault) private returns (uint256 weight) {
```
Comment (amandeeprehal-lab49) [inline contracts/incentive/Splitter.sol:130]: may be renamed to `vaultWeight`

File: contracts/incentive/Splitter.sol [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    function _calculateTokenRewardPercentage(IVault vault, uint256 totalWeight)
+        private
+        returns (uint256)
+    {
+        uint256 perShareReward = divide(_weightForVault(vault), totalWeight);
+        return perShareReward;
+    }
+
+    function _totalWeightForAllVaults() private returns (uint256 totalWeight) {
```
Comment (amandeeprehal-lab49) [inline contracts/incentive/Splitter.sol:122]: May be renamed to `totalVaultWeight`

File: contracts/incentive/Splitter.sol [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            totalWeightForAllVaults
+        );
+        amountToTransfer = multiply(totalAmount, percentage);
+    }
+
+    function _calculateTokenRewardPercentage(IVault vault, uint256 totalWeight)
+        private
+        returns (uint256)
+    {
+        uint256 perShareReward = divide(_weightForVault(vault), totalWeight);
```
Comment (amandeeprehal-lab49) [inline contracts/incentive/Splitter.sol:118]: I think, it's a particular vaultShare. perShareReward does not look correct

File: contracts/incentive/Vault.sol [hunk 6]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.5.0 <0.9.0;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "../common/hedera/HederaResponseCodes.sol";
+import "../common/IBaseHTS.sol";
+import "./IVault.sol";
+
+contract Vault is IVault, HederaResponseCodes, Initializable {
+    event VaultAdded(IVault vault, uint256 multiplier);
```
Comment (amandeeprehal-lab49) [inline contracts/incentive/Vault.sol:10]: I think, we are not using it so can be removed.

Comment (abhishek-k-sharma-lab49) [reply contracts/incentive/Vault.sol:10]: @amandeeprehal-lab49 Sir, it is being emitted from _addVault function.

File: contracts/incentive/IVault.sol (hunk 2 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/incentive/IVault.sol]: removed

File: contracts/incentive/Splitter.sol (hunk 4 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/incentive/Splitter.sol]: done

File: contracts/incentive/Splitter.sol (hunk 5 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/incentive/Splitter.sol]: done

File: contracts/incentive/Vault.sol (hunk 6 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/incentive/Vault.sol]: done

**Review by amandeeprehal-lab49 — APPROVED** [review]

**Pull request merged**